### DISCUSSION TITLE
How to set and get context.WorkflowInstance.Output of different activities in method from where workflow is started

### DISCUSSION THREAD

**Discussion by NiveditaGalwade** [elsa-workflows/elsa-core#5244] [Q&A]
public async Task<RunWorkflowResult> StartWorkflow<TRequest>(TRequest request, RunWorkflowResult workFlowResult, WorkflowInput workFlowInputItems)
      where TRequest : IWorkflowRequest
 {
     var workFlowBluePrint = await _workflowRegistry.FindAsync(request.WorkflowName, VersionOptions.Latest, null, CancellationToken.None);
     workFlowResult = workFlowBluePrint != null ? await _startsWorkflow.StartWorkflowAsync(workFlowBluePrint, null, workFlowInputItems) : null;
     return workFlowResult;
 }
 
 this is common method to intiate different types of workflow having multiple activities.
 How to set output (which is different type of objects) of each activity in workFlowResult at each activity 
 and then get those assigned output in above code.

**Comment by glime-ai** [comment]
To set the output of each activity in a workflow and then retrieve those outputs, you can use the `WorkflowExecutionContext` object's `Output` property within each activity's execution context. Here's a general approach to achieve this:

1. **Define the Output in Each Activity**: Within each activity's execution logic, you can set the output by assigning it to the `Output` property of the `WorkflowExecutionContext`. For example:

```csharp
public class MyActivity : Activity
{
    protected override async ValueTask<IActivityExecutionResult> OnExecuteAsync(ActivityExecutionContext context)
    {
        // Perform the activity's work here and determine the output
        var myOutput = new MyOutputType(); // Replace with actual output type and value

        // Set the output
        context.WorkflowExecutionContext.Output["MyActivityOutput"] = myOutput;

        return Done();
    }
}
```

2. **Retrieve Outputs After Workflow Execution**: After the workflow has been executed, you can access the outputs from the `WorkflowExecutionContext` object's `Output` property. Since you have a common method to initiate different types of workflows, you can retrieve the outputs as follows:

```csharp
public async Task<RunWorkflowResult> StartWorkflow<TRequest>(TRequest request, RunWorkflowResult workFlowResult, WorkflowInput workFlowInputItems)
    where TRequest : IWorkflowRequest
{
    var workFlowBluePrint = await _workflowRegistry.FindAsync(request.WorkflowName, VersionOptions.Latest, null, CancellationToken.None);
    if (workFlowBluePrint != null)
    {
        workFlowResult = await _startsWorkflow.StartWorkflowAsync(workFlowBluePrint, null, workFlowInputItems);

        // Assuming workFlowResult contains the WorkflowExecutionContext
        var outputs = workFlowResult.WorkflowExecutionContext.Output;

        // Access the output of a specific activity
        var myActivityOutput = outputs["MyActivityOutput"] as MyOutputType; // Replace with the actual key and type
    }
    return workFlowResult;
}
```

Please note that the above code assumes that `workFlowResult` contains the `WorkflowExecutionContext` after the workflow execution. If `workFlowResult` is of a different type that does not directly expose the `WorkflowExecutionContext`, you will need to adapt the code to access the `Output` property accordingly.

Remember to replace `MyOutputType` with the actual type of the output you expect from the activity, and `"MyActivityOutput"` with the actual key you used to store the output in the `Output` dictionary.

If you need further clarification or if the `RunWorkflowResult` type does not expose the `WorkflowExecutionContext` directly, please provide more details about the `RunWorkflowResult` type and how it is structured.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by NiveditaGalwade** [reply]
Severity	Code	Description	Project	File	Line	Suppression State
Error	CS1061	'WorkflowExecutionContext' does not contain a definition for 'Output' and no accessible extension method 'Output' accepting a first argument of type 'WorkflowExecutionContext' could be found (are you missing a using directive or an assembly reference?)

**Reply by NiveditaGalwade** [reply]
Can someone please address this query?

**Comment by NiveditaGalwade** [comment]
context.WorkflowExecutionContext.Output["MyActivityOutput"] = myOutput;
 
not able to set Output using above method. PLease send full workflow code with output set.

**Comment by NiveditaGalwade** [comment]
@sfmskywalker can you please suggest me solution for this query?
